File: src/main/java/org/cubexmc/metro/persistence/SaveCoordinator.java
package org.cubexmc.metro.persistence;

import java.io.IOException;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.util.ArrayList;
import java.util.Map;
import java.util.Objects;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.Executor;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Supplier;
import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * Coordinates persistent YAML saves so newer snapshots cannot be overwritten by
 * older async writes.
 */
public class SaveCoordinator {

    private static final int MAX_WRITE_ATTEMPTS = 2;

    private final Logger logger;
    private final Executor executor;
    private final Map<Path, AtomicLong> latestVersions = new ConcurrentHashMap<>();
    private final Map<Path, CompletableFuture<Void>> writeChains = new ConcurrentHashMap<>();

    public SaveCoordinator(Logger logger, Executor executor) {
        this.logger = Objects.requireNonNull(logger, "logger");
        this.executor = Objects.requireNonNull(executor, "executor");
    }

    public long submit(Path targetFile, Supplier<String> snapshotSupplier) {
        Objects.requireNonNull(snapshotSupplier, "snapshotSupplier");
        String snapshot = snapshotSupplier.get();
        return submitSnapshot(targetFile, snapshot);
    }

    public long submitSnapshot(Path targetFile, String snapshot) {
        Path normalizedTarget = normalize(targetFile);
        long version = nextVersion(normalizedTarget);

        writeChains.compute(normalizedTarget, (path, previous) -> {
            CompletableFuture<Void> base = previous == null ? CompletableFuture.completedFuture(null) : previous;
            CompletableFuture<Void> next = base.handle((ignored, previousError) -> null)
                    .thenRunAsync(() -> writeIfCurrent(path, version, snapshot), executor);
            next.whenComplete((ignored, error) -> {
                if (error != null) {
                    logger.log(Level.SEVERE, "Failed to save " + path + " at version " + version, error);
                }
                writeChains.remove(path, next);
            });
            return next;
        });

        return version;
    }

    public void saveNow(Path targetFile, String snapshot) {
        Path normalizedTarget = normalize(targetFile);
        flush(normalizedTarget);
        long version = nextVersion(normalizedTarget);
        writeSnapshot(normalizedTarget, version, snapshot);
    }

    public void flush(Path targetFile) {
        Path normalizedTarget = normalize(targetFile);
        CompletableFuture<Void> pending = writeChains.get(normalizedTarget);
        if (pending == null) {
            return;
        }
        pending.join();
    }

    public void flushAll() {
        for (CompletableFuture<Void> pending : new ArrayList<>(writeChains.values())) {
            pending.join();
        }
    }

    long latestVersion(Path targetFile) {
        AtomicLong version = latestVersions.get(normalize(targetFile));
        return version == null ? 0L : version.get();
    }

    private long nextVersion(Path targetFile) {
        return latestVersions.computeIfAbsent(targetFile, path -> new AtomicLong()).incrementAndGet();
    }

    private void writeIfCurrent(Path targetFile, long version, String snapshot) {
        if (!isCurrent(targetFile, version)) {
            logger.fine("Skipping stale save for " + targetFile + " at version " + version);
            return;
        }
        writeSnapshot(targetFile, version, snapshot);
    }

    private void writeSnapshot(Path targetFile, long version, String snapshot) {
        IOException lastError = null;
        for (int attempt = 1; attempt <= MAX_WRITE_ATTEMPTS; attempt++) {
            try {
                writeAtomically(targetFile, version, snapshot);
                return;
            } catch (IOException ex) {
                lastError = ex;
                logger.log(Level.WARNING, "Failed to save " + targetFile + " at version " + version
                        + " (attempt " + attempt + "/" + MAX_WRITE_ATTEMPTS + ")", ex);
            }
        }

        if (lastError != null) {
            logger.log(Level.SEVERE, "Giving up saving " + targetFile + " at version " + version, lastError);
        }
    }

    private void writeAtomically(Path targetFile, long version, String snapshot) throws IOException {
        Path parent = targetFile.getParent();
        if (parent != null) {
            Files.createDirectories(parent);
        }

        Path tempFile = targetFile.resolveSibling(targetFile.getFileName() + ".tmp-" + version);
        try {
            Files.writeString(tempFile, snapshot);
            if (!isCurrent(targetFile, version)) {
                Files.deleteIfExists(tempFile);
                logger.fine("Discarded stale save for " + targetFile + " at version " + version);
                return;
            }

            try {
                Files.move(tempFile, targetFile, StandardCopyOption.ATOMIC_MOVE,
                        StandardCopyOption.REPLACE_EXISTING);
            } catch (AtomicMoveNotSupportedException unsupported) {
                logger.fine("Atomic move is not supported for " + targetFile + "; falling back to replace existing.");
                Files.move(tempFile, targetFile, StandardCopyOption.REPLACE_EXISTING);
            }
        } finally {
            Files.deleteIfExists(tempFile);
        }
    }

    private boolean isCurrent(Path targetFile, long version) {
        return latestVersion(targetFile) == version;
    }

    private Path normalize(Path targetFile) {
        return Objects.requireNonNull(targetFile, "targetFile").toAbsolutePath().normalize();
    }
}

File: src/main/java/org/cubexmc/metro/persistence/SaveCoordinator.kt
package org.cubexmc.metro.persistence

import java.io.IOException
import java.nio.file.AtomicMoveNotSupportedException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.util.Objects
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.Executor
import java.util.concurrent.atomic.AtomicLong
import java.util.function.Supplier
import java.util.logging.Level
import java.util.logging.Logger

/**
 * Coordinates persistent YAML saves so newer snapshots cannot be overwritten by
 * older async writes.
 */
class SaveCoordinator(
    private val logger: Logger,
    private val executor: Executor,
) {
    private val latestVersions: MutableMap<Path, AtomicLong> = ConcurrentHashMap()
    private val writeChains: MutableMap<Path, CompletableFuture<Void>> = ConcurrentHashMap()

    init {
        Objects.requireNonNull(logger, "logger")
        Objects.requireNonNull(executor, "executor")
    }

    fun submit(targetFile: Path, snapshotSupplier: Supplier<String>): Long {
        Objects.requireNonNull(snapshotSupplier, "snapshotSupplier")
        val snapshot = snapshotSupplier.get()
        return submitSnapshot(targetFile, snapshot)
    }

    fun submitSnapshot(targetFile: Path, snapshot: String): Long {
        val normalizedTarget = normalize(targetFile)
        val version = nextVersion(normalizedTarget)

        writeChains.compute(normalizedTarget) { path, previous ->
            val base = previous ?: CompletableFuture.completedFuture(null)
            val next = base.handle { _, _ -> null }
                .thenRunAsync({ writeIfCurrent(path, version, snapshot) }, executor)
            next.whenComplete { _, error ->
                if (error != null) {
                    logger.log(Level.SEVERE, "Failed to save $path at version $version", error)
                }
                writeChains.remove(path, next)
            }
            next
        }

        return version
    }

    fun saveNow(targetFile: Path, snapshot: String) {
        val normalizedTarget = normalize(targetFile)
        flush(normalizedTarget)
        val version = nextVersion(normalizedTarget)
        writeSnapshot(normalizedTarget, version, snapshot)
    }

    fun flush(targetFile: Path) {
        val normalizedTarget = normalize(targetFile)
        val pending = writeChains[normalizedTarget] ?: return
        pending.join()
    }

    fun flushAll() {
        for (pending in ArrayList(writeChains.values)) {
            pending.join()
        }
    }

    fun latestVersion(targetFile: Path): Long {
        val version = latestVersions[normalize(targetFile)]
        return version?.get() ?: 0L
    }

    private fun nextVersion(targetFile: Path): Long =
        latestVersions.computeIfAbsent(targetFile) { AtomicLong() }.incrementAndGet()

    private fun writeIfCurrent(targetFile: Path, version: Long, snapshot: String) {
        if (!isCurrent(targetFile, version)) {
            logger.fine("Skipping stale save for $targetFile at version $version")
            return
        }
        writeSnapshot(targetFile, version, snapshot)
    }

    private fun writeSnapshot(targetFile: Path, version: Long, snapshot: String) {
        var lastError: IOException? = null
        for (attempt in 1..MAX_WRITE_ATTEMPTS) {
            try {
                writeAtomically(targetFile, version, snapshot)
                return
            } catch (ex: IOException) {
                lastError = ex
                logger.log(Level.WARNING, "Failed to save $targetFile at version $version (attempt $attempt/$MAX_WRITE_ATTEMPTS)", ex)
            }
        }

        if (lastError != null) {
            logger.log(Level.SEVERE, "Giving up saving $targetFile at version $version", lastError)
        }
    }

    private fun writeAtomically(targetFile: Path, version: Long, snapshot: String) {
        val parent = targetFile.parent
        if (parent != null) {
            Files.createDirectories(parent)
        }

        val tempFile = targetFile.resolveSibling(targetFile.fileName.toString() + ".tmp-" + version)
        try {
            Files.writeString(tempFile, snapshot)
            if (!isCurrent(targetFile, version)) {
                Files.deleteIfExists(tempFile)
                logger.fine("Discarded stale save for $targetFile at version $version")
                return
            }

            try {
                Files.move(tempFile, targetFile, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
            } catch (_: AtomicMoveNotSupportedException) {
                logger.fine("Atomic move is not supported for $targetFile; falling back to replace existing.")
                Files.move(tempFile, targetFile, StandardCopyOption.REPLACE_EXISTING)
            }
        } finally {
            Files.deleteIfExists(tempFile)
        }
    }

    private fun isCurrent(targetFile: Path, version: Long): Boolean = latestVersion(targetFile) == version

    private fun normalize(targetFile: Path): Path = Objects.requireNonNull(targetFile, "targetFile").toAbsolutePath().normalize()

    companion object {
        private const val MAX_WRITE_ATTEMPTS = 2
    }
}
